### ISSUE TITLE
Interaction between AllowNull and MemberNotNull raises a false warning when object initializer is used

### ISSUE THREAD

**Issue by algirdasN** [dotnet/roslyn#81594]
When a non-nullable property with AllowNull is initialized to null value, compiler raises a warning when attempting to set a field marked with MemberNotNull from that property. If the property is set to null after initialization, no warning is raised. Sorry if this description is confusing, the .NET Lab example should be self explanatory.

**Version Used**: 
.NET 10.0.100

**Steps to Reproduce**:

1. Open [Example](https://lab.razor.fyi/#nVHNSsNAELaiUvckPXoaemouIT2XWEoLomgpWBQU0W0yjYub3bo_raX05Bv4Bj6TL-ATePIBJElTarBSnNt8PzvfzpCXPUJ6SkaKxm6gK1-7VjMRwcVUG4wbZLVz25JzDAyTQrvHKFCx4A_FSRxbQwccC5oOo5GQ2rBAr2fctgyxJSifalaUnTHxVID6DwppyES0Dnf7VD_qBiEBp1pDRwY2RmHIjAAAjBQbU4OgjWIiasKdoDE2SMrdnGM8QNWVpms5ryWMHNZShePcZnY74CyAsWQhXFLOwpqT4tnjSY2pghEqLQX4IHACvbSpOY2lJKPdLo0x0VjOFwGSSseBvyrKnPN_hbyiSjARbRoTZrASC-abBZvny84e-XXVS381xCG13FTzrbc4l5PkMz_SL1zJmEL0CA34R_nlclSvoODDmHKL0FweOA9a2Trcqpbrnuu5dc873Q709X759fP97WN4sHNfei59Aw)
2. Compile

**Diagnostic Id**:

Warning CS8774 : Member '_name' must have a non-null value when exiting.

**Expected Behavior**:  No warning is raised.

**Actual Behavior**: Warning CS8774 is raised.

**Comment by jjonescz** [comment]
For future reference, this is the full code snippet:

```cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

class Document
{
    private string? _name;

    [MemberNotNull(nameof(_name))]
    public void Valid()
    {
        var person = new Person();
        person.Name = null;

        _name = person.Name;
    }

    [MemberNotNull(nameof(_name))]
    public void Warning()
    {
        var person = new Person { Name = null };

        _name = person.Name;
    }
}

class Person
{
    private string _name = "default";

    [AllowNull]
    public string Name
    {
        get => _name;
        set => _name = value ?? _name;
    }
}
```

Observed:

```cs
// (29,5): warning CS8774: Member '_name' must have a non-null value when exiting.
//     }
Diagnostic(ErrorCode.WRN_MemberNotNull, "}").WithArguments("_name").WithLocation(29, 5)
```

**Comment by jjonescz** [comment]
@RikkiGibson this inconsistency looks weird, shouldn't the warning either be reported in both methods (`Document.Valid` and `Document.Warning`) or none of them?
